src/main/java/org/openstreetmap/atlas/geography/atlas/change/diff/AtlasDiff.java

```diff
@@ -0,0 +1,467 @@
+     * Create a set of modify {@link FeatureChange}s ie. ADDs that change an existing feature. The
+     * set will contain an individual {@link FeatureChange} for certain types of differences. Eg. if
+     * a Point's location AND tags changed, then the set will contain two feature changes, one for
+     * the location and one for the tags. However, if a Node's location and connected edges change,
+     * this will come in as one feature change.<br>
+     *
+     * @param entity
+     * @param beforeAtlas
+     * @param afterAtlas
+     * @param useGeometryMatching
+     * @param useBloatedEntities
+     * @param saveAllGeometries
+     * @return a {@link Set} containing the possibly constructed {@link FeatureChange}s
+     */
+    private Set<FeatureChange> createModifyFeatureChanges(final AtlasEntity entity,
+            final Atlas beforeAtlas, final Atlas afterAtlas, final boolean useGeometryMatching,
+            final boolean useBloatedEntities, final boolean saveAllGeometries)
+    {
+        final Set<FeatureChange> featureChanges = new HashSet<>();
+
+        final AtlasEntity beforeEntity = entity.getType().entityForIdentifier(this.before,
+                entity.getIdentifier());
+        final AtlasEntity afterEntity = entity.getType().entityForIdentifier(this.after,
+                entity.getIdentifier());
+
+        if (beforeEntity == null || afterEntity == null)
+        {
+            throw new CoreException("Unexpected null entity. This should never happen.");
+        }
+
+        /*
+         * Determine if the entities are actually different in any relevant way. If so, then we can
+         * decide how to create the feature change.
+         */
+
+        /*
+         * Detect changed tags.
+         */
+        AtlasDiffHelper.getTagChangeIfNecessary(beforeEntity, afterEntity, useBloatedEntities,
+                saveAllGeometries).ifPresent(featureChanges::add);
+
+        /*
+         * Detect if the entity changed its parent relation membership.
+         */
+        AtlasDiffHelper
+                .getParentRelationMembershipChangeIfNecessary(beforeEntity, afterEntity,
+                        beforeAtlas, afterAtlas, useBloatedEntities, saveAllGeometries)
+                .ifPresent(featureChanges::add);
+
+        /*
+         * Detect if the entities were Nodes and some Node properties changed. We check for Node
+         * locations, as well as inEdges and outEdges.
+         */
+        if (entity instanceof Node)
+        {
+            AtlasDiffHelper
+                    .getNodeChangeIfNecessary((Node) beforeEntity, (Node) afterEntity,
+                            useGeometryMatching, useBloatedEntities, saveAllGeometries)
+                    .ifPresent(featureChanges::add);
+        }
+
+        /*
+         * Detect if the entities were Edges and some Edge properties changed. We check for Edge
+         * polylines, as well as the start and end Node.
+         */
+        if (entity instanceof Edge)
```

Comment:
else if? So that it does not re-evaluate every time?

Reply:
Yeah good catch. I was copy-pasting from the catch-all `getTagChanges`/`getParentRelationChanges` part and forgot to add the `else` statements here.